cirq-google/cirq_google/engine/stream_manager.py

```diff
@@ -59,5 +80,222 @@ def publish_exception(self, exception: GoogleAPICallError) -> None:
+                async for response in response_iterable:
+                    self._response_demux.publish(response)
+            except BaseException as e:
+                # TODO(#5996) Close the request iterator to close the existing stream.
+                # Note: the message ID counter is not reset upon a new stream.
+                self._response_demux.publish_exception(e)  # Raise to all request tasks
+                if isinstance(e, asyncio.CancelledError):
+                    break
+
+    async def _manage_execution(
+        self, project_name: str, program: quantum.QuantumProgram, job: quantum.QuantumJob
+    ) -> Union[quantum.QuantumResult, quantum.QuantumJob]:
+        """The execution coroutine, an asyncio coroutine to manage the lifecycle of a job execution.
+
+        This coroutine sends QuantumRunStream requests to the request iterator and receives
+        responses from the ResponseDemux.
+
+        It initially sends a CreateQuantumProgramAndJobRequest, and retries if there is a retryable
+        error by sending another request. The exact request type depends on the error.
+
+        There is one execution coroutine per running job submission.
+        """
+
+        # Construct requests ahead of time to be reused for retries.
+        create_program_and_job_request = quantum.QuantumRunStreamRequest(
+            parent=project_name,
+            create_quantum_program_and_job=quantum.CreateQuantumProgramAndJobRequest(
+                parent=project_name, quantum_program=program, quantum_job=job
+            ),
+        )
+        create_job_request = quantum.QuantumRunStreamRequest(
+            parent=project_name,
+            create_quantum_job=quantum.CreateQuantumJobRequest(
+                parent=program.name, quantum_job=job
+            ),
+        )
+        get_result_request = quantum.QuantumRunStreamRequest(
+            parent=project_name, get_quantum_result=quantum.GetQuantumResultRequest(parent=job.name)
+        )
+
+        current_request = create_program_and_job_request
+        while True:
+            try:
+                current_request.message_id = self._generate_message_id()
+                response_future = self._response_demux.subscribe(current_request.message_id)
+                await self._request_queue.put(current_request)
+                response = await response_future
+
+            # Broken stream
+            except google_exceptions.GoogleAPICallError as e:
+                if _is_retryable_error(e):
+                    # Retry
+                    self._response_demux.unsubscribe(current_request.message_id)
+                    current_request = get_result_request
+                    continue
+                    # TODO(#5996) add exponential backoff
+                raise e
+
+            # Either when this request is canceled or the _manage_stream() loop is canceled.
+            except asyncio.CancelledError:
+                if response_future is not None:
```

Comment:
...and not completed? Else you'll get another exception here if that case is tripped.

Reply:
Cancelling an already cancelled future returns False instead of raising. It would be good practice to check the future is not done, but it actually requires a bigger change that I don't think I should do now: if `response_future` is cancelled, it is also `done()`, and to fix this, it would be best to move the future cancellation logic into a callback instead of processing it here as an exception, but we need extra changes to somehow pass the job name into the callback.